web/web_test.go

```diff
@@ -0,0 +1,1868 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package web
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+	"github.com/cloudspannerecosystem/harbourbridge/spanner/ddl"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetTypeMapNoDriver(t *testing.T) {
+	req, err := http.NewRequest("GET", "/typemap", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	rr := httptest.NewRecorder()
+	handler := http.HandlerFunc(getTypeMap)
+	handler.ServeHTTP(rr, req)
+	if status := rr.Code; status != http.StatusNotFound {
+		t.Errorf("handler returned wrong status code: got %v want %v",
+			status, http.StatusNotFound)
+	}
+
+}
+
+func TestGetTypeMapPostgres(t *testing.T) {
+	app.driver = "postgres"
+	app.conv = internal.MakeConv()
+	buildConvPostgres(app.conv)
+	req, err := http.NewRequest("GET", "/typemap", nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	rr := httptest.NewRecorder()
+	handler := http.HandlerFunc(getTypeMap)
+	handler.ServeHTTP(rr, req)
+	var typemap map[string][]typeIssue
+	json.Unmarshal(rr.Body.Bytes(), &typemap)
+	if status := rr.Code; status != http.StatusOK {
+		t.Errorf("handler returned wrong status code: got %v want %v",
+			status, http.StatusOK)
+	}
+	expectedTypemap := map[string][]typeIssue{
```

Comment:
There isn't a lot of value in listing down the complete type map here since we are not validating the correctness of the type map itself. But I understand that the map as written down here doesn't exist in the code as a constant either. A middle ground here may be to only compare expectedTypeMap for a few types let's say bool and numeric. That way we will be able to test that marshalling etc and other pieces in the handler are working as intended.

Reply:
Similarly in TestGetTypeMapMysql

Reply:
i didn't understand this completely, are you saying that we should use less number of types in expectedTypemap?
note that typemap is depended on actual conv, we need to check whether it outputs all the types included in source database.

Reply:
Ah, I see. This makes sense then since you are also doing the filtering on API layer to only return the type map for types that are present in the conv struct. What confused me was the fact that buildConvPostgres contains almost all the types (if not all?). SGTM.